### ISSUE TITLE
Add support for Java 8 functional interfaces as alternatives to Gradle's

### ISSUE THREAD

**Issue by adammurdoch** [gradle/gradle#2846]
The Gradle API includes a number of functional interfaces that predate those added in Java 8, but which serve the same purpose. For example

- Gradle's `Action<T>` is the same in intent as `Consumer<T>`
- Gradle's `Transformer<OUT, IN>` is the same in intent as `Function<IN, OUT>`
- Gradle's `Spec<T>` is the same in intent as `Predicate<T>`.

Using Gradle-specific types has a few disadvantages:

1. I need to understand what they mean. I'm more likely to have seen the Java 8 interfaces before.
2. I cannot easily use implementations of these type provided by others, say from some library.

We should add API overloads and other support for these types, so that build scripts and plugins can use these types. Later, we should deprecate and remove the Gradle interfaces in preference to the Java 8 interfaces.

### Expected Behavior

I can, for example, use a `Predicate` implementation to exclude files from a `FileTree`, or use a `Consumer<JavaCompile>` instance to configure the `JavaCompile` tasks.

### Current Behavior

I have to adapt between Java type and Gradle types.

**Comment by ldaley** [comment]
One issue to sort out is that overloading methods with SAM types hurts usability in that you cannot use lambdas as parameters, without casting. 

Only reliable options I know of are:

1. Live with making users cast to use lambda syntax
2. Use different method names

It might be worth experimenting with replacing `Action` parameter types with `Consumer` etc. and making `Action` extend `Consumer`. This will be breaking binary change, but might be source compatible and allow lambda syntax going forward.

**Comment by JLLeitschuh** [comment]
The problem is that Gradle needs to work with JDK 7 so this won't be possible till gradle no longer supports projects being compiled using JDK 7.

**Comment by big-guy** [comment]
@JLLeitschuh We're going to deprecate _running_ Gradle with Java 7 soon, but you'll still be able to compile with JDK7 for a long time.

**Comment by JLLeitschuh** [comment]
Right, that's what I was trying to convey but doing a bad job of it. Thank you for clarifying.

**Comment by JLLeitschuh** [comment]
@adammurdoch now that we use JDK 8 internally, any reason why we couldn't just have `Action` extend `Consumer`, `Transformer` extend `Function`, ect...

**Comment by eskatos** [comment]
@JLLeitschuh that would be a breaking change

**Comment by JLLeitschuh** [comment]
I'm having a hard time wrapping my head around why it would be an API breaking change, can you elaborate?

**Comment by eskatos** [comment]
`Action` is a type implemented by user code. If you make it `extends Consumer` it requires a new method. Breaking user code. We could add a default method, but we can't as it needs to be compiled against Java 6 currently.
